### ISSUE TITLE
Allow manually specifying a (non-enforced) primary key on sources 

### ISSUE THREAD

**Issue by aljoscha** [MaterializeInc/materialize#6985]
Currently, Materialize only knows about primary keys for Kafka and Postgres sources. This possibly reduces optimization potential and in some cases prevents users from using certain constructs.

The immediate motivation for this is https://github.com/MaterializeInc/materialize/pull/6709 which added a check to UPSERT sinks that verifies that the sinked relation actually has a compatible primary key. This stricter check prevents use cases where input data has a known primary key but there is no way of telling the system.

I propose adding an optional `PRIMARY KEY (field1, [field2], ...) NOT ENFORCED` part to the `CREATE SOURCE ...` DDL. Specifically, in the optional part where column names can be specified. Alternatively, we could allow `PRIMARY KEY` right after the optional column names.

The `NOT ENFORCED` suffix is an important signal for users: Materialize will not enforce the constraint and will produce wrong results if it is not correct. We must also clearly document this. But, `NOT ENFORCED` seems to be common enough in other systems so the behavior should not be too surprising.

Concretely, we should allow something like this:

```
CREATE SOURCE the_source (a, b, PRIMARY KEY (a) NOT ENFORCED)
  FROM KAFKA BROKER '${testdrive.kafka-addr}'
  TOPIC 'topic'
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
  ENVELOPE NONE
```

Or, alternatively:

```
CREATE SOURCE the_source PRIMARY KEY (a) NOT ENFORCED
  FROM KAFKA BROKER '${testdrive.kafka-addr}'
  TOPIC 'topic'
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
  ENVELOPE NONE
```
Remains the question of whether the new part should be called `PRIMARY KEY` or `UNIQUE` and whether we use `NOT ENFORCED` or `UNENFORCED`, or some other alternative. A quick survey shows that most data base systems have both `UNIQUE` and `PRIMARY KEY` and both would be valid for our purposes. All surveyed systems use `NOT ENFORCED`. Flink uses the `PRIMARY KEY (foo, bar) NOT ENFORCED` syntax for exactly the proposed purpose.

References:
 - https://dev.mysql.com/doc/refman/8.0/en/create-table.html
 - Postgres doesn't seem to have the `NOT ENFORCED` concept
 - https://www.ibm.com/docs/en/db2/11.5?topic=statements-create-table
 - https://docs.microsoft.com/en-us/azure/synapse-analytics/sql-data-warehouse/sql-data-warehouse-table-constraints
 - https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/connectors/table/overview/#primary-key

**Comment by benesch** [comment]
This is a wonderful writeup! Thanks very much for doing the survey of the other database systems. I don't think we need any more of a design doc than this. Let's get this scheduled!

**Comment by umanwizard** [comment]
+1

**Comment by uce** [comment]
Fixed in #7088.

I'm closing this issue, but there is a potential follow-up to allow field expressions for selecting nested fields as part of the key. We should create a separate ticket for that if we plan to implement it.
